lib/extras/encode_jpeg.cc

```diff
@@ -0,0 +1,340 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include "lib/extras/encode_jpeg.h"
+
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "lib/extras/encode_jpeg.cc"
+#include <hwy/foreach_target.h>
+#include <hwy/highway.h>
+
+#include "lib/jxl/color_encoding_internal.h"
+#include "lib/jxl/common.h"
+#include "lib/jxl/enc_adaptive_quantization.h"
+#include "lib/jxl/enc_cluster.h"
+#include "lib/jxl/enc_color_management.h"
+#include "lib/jxl/enc_quant_weights.h"
+#include "lib/jxl/enc_transforms.h"
+#include "lib/jxl/enc_xyb.h"
+#include "lib/jxl/huffman_tree.h"
+#include "lib/jxl/image.h"
+#include "lib/jxl/jpeg/dec_jpeg_data_writer.h"
+#include "lib/jxl/jpeg/jpeg_data.h"
+#include "lib/jxl/opsin_params.h"
+HWY_BEFORE_NAMESPACE();
+namespace jxl {
+namespace extras {
+namespace HWY_NAMESPACE {
+
+void ComputeDCTCoefficients(const Image3F& opsin, const ImageF& qf,
+                            const FrameDimensions& frame_dim, const float* qm,
+                            std::vector<jpeg::JPEGComponent>* components) {
+  float qfmin, qfmax;
+  ImageMinMax(qf, &qfmin, &qfmax);
+  HWY_ALIGN float scratch_space[2 * kDCTBlockSize];
+  for (size_t c = 0; c < 3; c++) {
+    std::vector<jpeg::coeff_t>& coeffs = (*components)[c].coeffs;
+    size_t num_blocks = frame_dim.xsize_blocks * frame_dim.ysize_blocks;
+    coeffs.resize(num_blocks * kDCTBlockSize);
+    const float* qmc = &qm[c * kDCTBlockSize];
+    for (size_t by = 0, bix = 0; by < frame_dim.ysize_blocks; by++) {
+      for (size_t bx = 0; bx < frame_dim.xsize_blocks; bx++, bix++) {
+        HWY_ALIGN float dct[kDCTBlockSize];
+        TransformFromPixels(AcStrategy::Type::DCT,
+                            opsin.PlaneRow(c, 8 * by) + 8 * bx,
+                            opsin.PixelsPerRow(), dct, scratch_space);
+        jpeg::coeff_t* block = &coeffs[bix * kDCTBlockSize];
+        for (size_t iy = 0, i = 0; iy < 8; iy++) {
+          for (size_t ix = 0; ix < 8; ix++, i++) {
+            float coeff = 2040 * dct[i] * qmc[i];
+            float zero_bias = 0.5f * qfmax / qf.Row(by)[bx];
+            int cc = std::abs(coeff) < zero_bias ? 0 : std::round(coeff);
+            if (zero_bias > 1) {
```

Comment:
I probably wrote this, so I don't have a lot of grounds to complain :P but some comments on what is going on here wouldn't hurt :)

Reply:
Done.